src/pkcs11/misc.c

```diff
@@ -202,9 +201,11 @@ CK_RV push_login_state(struct sc_pkcs11_slot *slot,
 	r = CKR_OK;
 
 err:
-	if (r != CKR_OK && login) {
-		sc_mem_clear(login->pPin, login->ulPinLen);
-		free(login->pPin);
+	if (login) {
```

Comment:
`login` can be inside the `slot->logins` queue here. You could to the following to avoid this:
```
if (0 > list_append(&slot->logins, login)) {
		goto err;
	}
login = NULL;
```